File: Sample/Android/kotlin/MorningSDKSample/app/src/main/java/com/morningsoft/morningsdksample/WebAppInterface.kt
package com.morningsoft.morningsdksample

//
//  WebAppInterface.kt
//  MorningSDKSample
//
//  Created by Kyojun.Kim on 18/07/2023.
//  Copyright © 2023 Morningsoft. All rights reserved.
//

import android.annotation.SuppressLint
import android.content.Context
import android.os.Handler
import android.os.Looper
import android.webkit.JavascriptInterface
import android.webkit.WebView

class WebAppInterface(webView : WebView?, var context: Context?){
    private var lstPacket : MutableList<WebviewPacket> =  mutableListOf()

    init {
        mWebView = webView
        mContext = context
    }

    @JavascriptInterface
    fun postMessage(data: String) {
        val webviewPacket = WebviewPacket(data)
        lstPacket.add(webviewPacket)
        Handler(Looper.getMainLooper()).postDelayed({
            processPacket()
        }, 1000)
    }
    private fun processPacket() {
        if(lstPacket.size > 0) {
            val webviewPacket = lstPacket.removeLast()
            MainApplication().sendMessage(webviewPacket)
        }
        if(lstPacket.size > 0) {
            Handler(Looper.getMainLooper()).postDelayed({
                processPacket()
            }, 1000)
        }
    }
    fun sendComplete(webviewPacket : WebviewPacket) {
        var methodName = webviewPacket.getMethodName()
        var arguments = webviewPacket.getArguments()
        val callback = webviewPacket.getCallback()
        Handler(Looper.getMainLooper()).postDelayed({
            webviewPacket.setMethodName(methodName)
            webviewPacket.setArguments(arguments)
            webviewPacket.setCallback(callback)
            webviewPacket.setResultCode(MainApplication().result_success)
            mWebView?.loadUrl("javascript:receiveBase64('" + webviewPacket.toBase64JSON() + "')")
        }, 3000)
    }

    @SuppressLint("StaticFieldLeak")
    companion object {
        private const val LOG_TAG = "WebAppInterface"
        private var mWebView: WebView? = null
        private var mContext: Context? = null
    }
}

File: Sample/iOS/swift/MorningSDKSample/MorningSDKSample/WebAppInterface.swift
//
//  WebAppInterface.swift
//  MorningSDKSample
//
//  Created by Kyojun.Kim on 17/05/2023.
//  Copyright © 2023 Morningsoft. All rights reserved.
//

import Dispatch
import WebKit

class WebAppInterface {
    private var LOG_TAG = "WebAppInterface"
    private var lstPacket : Array<WebviewPacket> = Array()
    var mWebView : WKWebView!
    
    init(webview : WKWebView) {
        mWebView = webview
    }
    func processNative(packetstring: String?) {
        print(LOG_TAG + " " + "processNative : " + packetstring!)
        let webviewPacket = WebviewPacket(param: packetstring!)
        print(LOG_TAG + " " + "processNative : " + webviewPacket.toJSON())
        lstPacket.append(webviewPacket)
        let mainQueue = DispatchQueue.main
        let deadline = DispatchTime.now() + .seconds(1)
        mainQueue.asyncAfter(deadline: deadline) {
            self.processPacket()
        }
    }
    private func processPacket() {
        if(lstPacket.count > 0) {
            let webviewPacket = lstPacket.removeLast()
            print(LOG_TAG + " " + "processPacket : " + webviewPacket.toJSON())
            MainApplication().sendMessage(webviewPacket : webviewPacket)
        }
        if(lstPacket.count > 0) {
            let mainQueue = DispatchQueue.main
            let deadline = DispatchTime.now() + .seconds(1)
            mainQueue.asyncAfter(deadline: deadline) {
                self.processPacket()
            }
        }
    }
    func sendComplete(webviewPacket : WebviewPacket) {
        print(LOG_TAG + " " + "sendComplete : " + webviewPacket.toJSON())
        let arguments = webviewPacket.getArguments()
        let callback = webviewPacket.getCallback()
        let mainQueue = DispatchQueue.main
        let deadline = DispatchTime.now() + .seconds(1)
        mainQueue.asyncAfter(deadline: deadline) {
            webviewPacket.setMethodName(strMethodName: callback)
            webviewPacket.setArguments(strArguments: arguments)
            webviewPacket.setCallback(strCallback: callback)
            webviewPacket.setResultCode(strResultCode: MainApplication().RESULT_SUCCESS)
            let javascriptString = "receiveBase64('" + webviewPacket.toBase64JSON() + "');"
            self.mWebView.evaluateJavaScript(javascriptString, completionHandler: {(result, error) in
                if let result = result {
                    print(result)
                }
            })
        }
    }
}
